### PR TITLE
Script/Action for Issue 11

### LINKED ISSUE #11: Build a script to collect the github repos for CodeForPhilly
We need the list of github repos in the following format for the tap-github:

> Update - it is space-delimited, not comma-delimited

`repository: org/repo_name user/repo_name org/repo_name`

@themightychris has pointed out we can get the list of CFP projects from https://brigade.cloud/static-api/v1/organizations/Code%20for%20Philly/projects.json

Definition of done:
- [ ] Build a script (python or whatever) to iterate through the projects and output like above: `repository: CodeForPhilly/cfp-data-pipeline`
- [ ] Have the script update the `meltano.yml` file https://github.com/CodeForPhilly/cfp-data-pipeline/blob/master/meltano/cfp-pipeline/meltano.yml#L17 to update the `repository` line.
- [ ] Bonus - build a github action to run your script

### PR DESCRIPTION
**Pull request by jaredgk**
Here's what I have for this issue:
-Python script that parses JSON of CFP repo data, prints out a line for sed to use in replacing line in meltano.yml
-Github action yml that adds a user-triggered action to update the repository line in meltano.yml and push the changes to the master branch

It's functional but there's a few little things that might need to be changed, like adding the action into the main action yml, using a different directory for storing the python script, and a general question about the edit to the meltano file since there's over 200 repos added so the file is a little messy afterwards.

### CODE REVIEW

**Comment by machow** [comment]
@jaredgk (cc @dherbst @chriscardillo) thanks for working on this! It seems like all the pieces are there--the big piece I wonder about is how we should stitch them all together.

It seems like data pipelines have roughly two levels:

* configuration (held in e.g. a git repository)
* tasks that generate and consume data (e.g. airflow dag)

The PR enables us to pull data on github repos, but feeds it into the configuration. If we could have two tasks like (pull repo data) -> (run repo ingest tap), then we can avoid a tough situation: changing our configuration when the data changes.

### Proposal

We create a task that pulls the repo data in airflow, have the tap depend on that task.

* How easy is it to add task dependencies in meltano, or is it easier to wire up the tap manually?
* How to pass data to the tap? Could save to the warehouse, then set as environment variable before running tap, etc..

Curious on people's takes here :)

**Comment by chriscardillo** [comment]
I don't think we can have our cake and eat it, too, here. Here is my two cents.

`meltano.yml` must contain all information about taps and targets in order for Meltano to work. That's what we signed up for when we chose to use Meltano, a layer of abstraction around taps and targets. If we continue to use Meltano, we must be committed to keeping `meltano.yml` in source control.

The alternative would be to remove the layer of abstraction (read as: Meltano), make any necessary changes to this repository for utilizing taps and targets directly, and then move the repo scraping script in this review to the front of what is currently `dags/meltano.py`, at which point we can make tap-target tasks dynamically in what is currently the Meltano DAG.

I'd vote for keeping `meltano.yml` and `meltano.py` as is, utilizing the Python script for sure, and using the Github Action if we are comfortable with this action triggering our CD pipeline. I came to this conclusion for two reasons:

1. Refactoring the pipeline to use taps and targets directly seems like taking a step back, retracing our steps and some hard work. There are definitely other places we could focus our energy to achieve further progress in this project.

2. I, personally, think that every task that runs in the pipeline belongs explicitly in source control. The jump from 100 github repos ingested to 120 github repos ingested in the pipeline should not be determined by the addition of 20 github repos in the github organization, but instead by the additional entries of these repos into source control (e.g as entries in `meltano.yml`). This allows us to monitor what we ingest over time, and limits the amount of noise in the warehouse.

Happy to discuss further.

**Comment by dherbst** [comment]
I think this fits the definition of done for the issue.  I was hoping we would set this on a cron github action, but it could create a lot of commit churn, which is what I think @machow wants to avoid.

I don't think we have to abandon meltano to get what we want, @chriscardillo.   Since the bulk of this is in python, we could add this to the bash script to create a file the tap can optionally read the list of files from.  So the execution for tap-github cold look something like:

```
airflow --> bashcommand --> A line runs a python script to get the repos and save to a file --> tap-github reads the list of repos from the file.
```

I don't think we can make it a separate airflow task unless we have the task write the repo list somewhere the tap-github task can read from.  Maybe from snowflake?  But it seems easier to do the lookup in the tap-github bashcommand and read from a file which is not checked into git to avoid commit churn. 

I think down the road we might encounter cases where we need to avoid pulling from some of the git repos on the list, so it might be best to keep it in meltano.yml but we'll have to see if we get there.

File: scripts/parse_repo_json.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+repo_json = json.loads(str(sys.stdin.readline()))
+
+out_string = ''
+for repo in repo_json['data']:
+    if repo['type'] == 'project':
+        repo_name = repo['id'].replace(' ','')
+        out_string += repo_name+' '
+
+print (out_string)
```
Comment (dherbst) [inline scripts/parse_repo_json.py:15]: Consider adding `strip()` to remove a trailing space, or avoid adding an extra space. 
```suggestion
print(out_string.strip())
```

File: scripts/parse_repo_json.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import sys
+import json
+
+
```
Comment (dherbst) [inline scripts/parse_repo_json.py:4]: You might want to consider running this file through https://pypi.org/project/pycodestyle/

**Review by dherbst — APPROVED** [review]
A couple of stylistic things you might want to consider, but looks good to me.  Thanks!

**Comment by machow** [comment]
Thanks for taking a look @chriscardillo and @dherbst. It sounds like both of you agree that we might as well check it in to the config. **I'm on board, since I'd rather have data in the pipe :p.**

## Future music: meltano task dependencies seem very doable

I wonder if there's a quick technical path to capture both sentiments:

1. Chris: getting away from meltano would be a step backward
2. Darrel: we could run the repo-fetching script before, then have the tap read it (e.g. from a file)

It looks like our pipeline uses meltano in a simple way--as BashOperator tasks, running e.g...

```
{meltano_bin} elt tap-github target-postgres-github --job_id=github-to-postgres
```

What's keeping us from adding dependencies to these tasks? It seems like we could...

1. Add each task to a dictionary, rather than re-assigning the variable elt [in this for loop](https://github.com/CodeForPhilly/cfp-data-pipeline/blob/master/airflow/dags/meltano.py#L44).
2. Add this repo-fetching script as a task that produces a csv.
3. Specify that the meltano github tasks depend on it (e.g. reference from the dict in step 1).

**Pull request merged**